subsys/bluetooth/mesh/access.c

```diff
@@ -2443,187 +2581,231 @@ static int read_comp_cb(const char *key, size_t len, settings_read_cb read_cb,
+	err = settings_load_subtree_direct(path, stored_page_read_cb, &read_buf);
 
 	if (err) {
-		LOG_ERR("Failed reading composition data: %d", err);
+		LOG_ERR("Failed reading data page: %d", err);
 		return err;
 	}
-	if (buf->len == original_len) {
+
+	if (read_buf.len == 0) {
 		return -ENOENT;
 	}
-	return 0;
-}
 
-int bt_mesh_model_data_store(const struct bt_mesh_model *mod, bool vnd,
-			     const char *name, const void *data,
-			     size_t data_len)
-{
-	char path[30];
-	int err;
-
-	encode_mod_path(mod, vnd, "data", path, sizeof(path));
-	if (name) {
-		strcat(path, "/");
-		strncat(path, name, SETTINGS_MAX_DIR_DEPTH);
+	if (read_buf.len == 1 && read_buf.data[0] == 0) {
+		/* Single 0 byte is a sentinel value for empty page, return
+		 * success without writing any bytes to the buffer.
+		 */
+		return 0;
 	}
 
-	if (data_len) {
-		err = settings_save_one(path, data, data_len);
-	} else {
-		err = settings_delete(path);
+	if (offset > read_buf.len) {
+		return 0;
 	}
 
-	if (err) {
-		LOG_ERR("Failed to store %s value", path);
-	} else {
-		LOG_DBG("Stored %s value", path);
+	if (!split_elems) {
+		return write_cdp_elems(buf, &read_buf, page);
 	}
-	return err;
+
+	len = MIN(net_buf_simple_tailroom(buf), read_buf.len - offset);
+	net_buf_simple_add_mem(buf, read_buf.data + offset, len);
+
+	return 0;
 }
 
-#if defined(CONFIG_BT_MESH_LARGE_COMP_DATA_SRV)
-static int metadata_set(const char *name, size_t len_rd, settings_read_cb read_cb, void *cb_arg)
+static int stored_page_size_cb(const char *key, size_t len, settings_read_cb read_cb, void *cb_arg,
+			       void *param)
 {
-	/* Only need to know that the entry exists. Will load the contents on
-	 * demand.
-	 */
-	if (len_rd > 0) {
-		atomic_set_bit(bt_mesh.flags, BT_MESH_METADATA_DIRTY);
+	size_t *size = param;
+
+	if (len > 0) {
+		*size = len;
 	}
 
 	return 0;
 }
-BT_MESH_SETTINGS_DEFINE(metadata, "metadata", metadata_set);
 
-int bt_mesh_models_metadata_store(void)
+static size_t stored_page_size_get(enum page_type type, uint8_t page)
 {
-	NET_BUF_SIMPLE_DEFINE(buf, CONFIG_BT_MESH_MODELS_METADATA_PAGE_LEN);
-	size_t total_size;
 	int err;
+	size_t size = 0;
+	const char *path;
 
-	total_size = bt_mesh_metadata_page_0_size();
-	LOG_DBG("bt/mesh/metadata total %d", total_size);
-
-	net_buf_simple_init(&buf, 0);
-	net_buf_simple_add_le16(&buf, total_size);
+	path = stored_page_path(type, page);
 
-	err = bt_mesh_metadata_get_page_0(&buf, 0);
-	if (err == -E2BIG) {
-		LOG_ERR("Metadata too large");
-		return err;
-	}
-	if (err) {
-		LOG_ERR("Failed to read models metadata: %d", err);
-		return err;
+	if (path == NULL) {
+		return 0;
 	}
 
-	LOG_DBG("bt/mesh/metadata len %d", buf.len);
-
-	err = settings_save_one("bt/mesh/metadata", buf.data, buf.len);
+	err = settings_load_subtree_direct(path, stored_page_size_cb, &size);
 	if (err) {
-		LOG_ERR("Failed to store models metadata: %d", err);
-	} else {
-		LOG_DBG("Stored models metadata");
+		return 0;
```

Comment:
Same as a previous comment, why not return `err`?

Reply:
I'd have to change the type to `ssize_t` everywhere then, including up to the `access.h` APIs. I didn't see the need to do this, since "page storage has some kind of error" and "Page is not there" is, from the perspective of the models trying to read the page, the same: It will have to report 0 length to a client.

Reply:
True, what about adding a `LOG_ERR(...)` message instead before returning 0?